src/opentimelineio/editAlgorithm.cpp

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Contributors to the OpenTimelineIO project
+
+#include "opentimelineio/editAlgorithm.h"
+
+#include "opentimelineio/gap.h"
+
+namespace opentimelineio { namespace OPENTIMELINEIO_VERSION {
+
+void
+edit_slice(
+    Composition*        composition,
+    RationalTime const& time,
+    ErrorStatus*        error_status)
+{
+    // Find the item to slice.
+    auto item = dynamic_retainer_cast<Item>(
+        composition->child_at_time(time, error_status, true));
+    if (!item)
+    {
+        return;
+    }
+    const int       index = composition->index_of_child(item);
+    const TimeRange range = composition->trimmed_range_of_child_at_index(index);
+
+    // Adjust the source range for the first slice.
+    const TimeRange first_source_range(
+        item->trimmed_range().start_time(),
+        time - range.start_time());
+    item->set_source_range(first_source_range);
+
+    // Clone the item for the second slice.
+    auto            second_item = dynamic_cast<Item*>(item->clone());
+    const TimeRange second_source_range(
+        first_source_range.start_time() + first_source_range.duration(),
+        range.duration() - first_source_range.duration());
+    second_item->set_source_range(second_source_range);
+    composition->insert_child(static_cast<int>(index) + 1, second_item);
```

Comment:
We should think about how to handle trimming Transitions that might overlap the slice. There is/was an issue in the flattening code which dealt with this. I think we ended up truncating or removing any Transition that overlapped a trim? Some day we will need to add a keyframed transition parameter or some other mechanism for modeling truncated transitions properly, but recommend keeping that out of the scope of this PR.

Also, after trimming the start/end of an item, it would be nice to remove any markers that are outside the new range.

The same things apply for any edit operation, not just slice.

Reply:
It seems like there are a number of tricky cases involving transitions, I wonder if the design doc needs to be updated with how to handle them. That's also a good point about the markers.